shared_crates/ui/src/node.rs

```diff
@@ -0,0 +1,400 @@
+//! Defines the [Node] element.
+
+use crate::prelude::*;
+use ambient_element::{element_component, to_owned, Element, ElementComponent, Hooks};
+use ambient_guest_bridge::{
+    components::{
+        app::cursor_position,
+        ecs::parent,
+        input::{mouse_over, mouse_pickable_max, mouse_pickable_min},
+        layout::{
+            align_vertical_center, fit_horizontal_parent, height, margin, min_height, padding,
+            space_between_items, width,
+        },
+        rect::{
+            background_color, border_color, border_radius, border_thickness, line_from, line_to,
+            line_width,
+        },
+        text::font_style,
+        transform::{local_to_parent, local_to_world, translation},
+    },
+    messages,
+};
+use ambient_shared_types::{CursorIcon, ModifiersState, VirtualKeyCode};
+use glam::{vec2, vec3, vec4, Mat4, Vec2, Vec3, Vec4};
+use std::str::FromStr;
+
+#[element_component]
+/// A Graph Element that you can put lots of nodes
+pub fn Graph(hooks: &mut Hooks) -> Element {
+    let (nodes, set_nodes) = hooks.use_state(vec![]);
+    let in_id = hooks.use_ref_with(|_| vec![]);
+    let out_id = hooks.use_ref_with(|_| vec![]);
+    let start_id = hooks.use_ref_with(|_| None);
+    let (lines, set_lines) = hooks.use_state(vec![]);
+    let (temp_line, set_temp_line) = hooks.use_state(None::<(Vec3, Vec3)>);
+    let temp_line_toggle = hooks.use_ref_with(|_| false);
+    hooks.use_frame({
+        to_owned![lines, set_lines, temp_line, set_temp_line, temp_line_toggle];
+        move |world| {
+            if *temp_line_toggle.lock() {
+                if let Some(line) = temp_line {
+                    let mouse_pos = world.resource(cursor_position());
+                    set_temp_line(Some((line.0, mouse_pos.extend(line.1.z))));
+                }
+            }
+
+            for (index, ((from_id, from_pos), (to_id, to_pos))) in lines.iter().enumerate() {
+                // let new_from_pos = world.get(*from_id, translation()).unwrap_or(Vec3::ZERO);
+                // let new_to_pos = world.get(*to_id, translation()).unwrap_or(Vec3::ZERO);
+                let ltw = world.get(*from_id, local_to_world()).unwrap();
+                let (_, _, new_from_pos) = Mat4::to_scale_rotation_translation(&ltw);
+                let new_from_pos = new_from_pos + vec3(5.0, 5.0, 0.0);
+
+                if new_from_pos != *from_pos {
+                    let mut l = lines.clone();
+                    l[index] = ((*from_id, new_from_pos), (*to_id, *to_pos));
+                    set_lines(l);
+                }
+
+                let ltw = world.get(*to_id, local_to_world()).unwrap();
+                let (_, _, new_to_pos) = Mat4::to_scale_rotation_translation(&ltw);
```

Comment:
I used `translation()` before but this will require its parent's `translation()` as well. The `DragArea` is the parent while the main body and the dots are children.